src/react/components/pages/editorPage/editorPage.tsx

```diff
@@ -484,9 +542,103 @@ export default class EditorPage extends React.Component<IEditorPageProps, IEdito
+            }
         }
     }
 
+    private predictImage = async (image: HTMLImageElement) => {
+        const regions = [...this.state.selectedAsset.regions];
+
+        const xRatio = this.state.selectedAsset.asset.size.width / image.width;
+        const yRatio = this.state.selectedAsset.asset.size.height / image.height;
+
+        const predictions = await this.model.detect(image);
+        predictions.forEach((prediction) => {
+            const left = Math.max(0, prediction.bbox[0] * xRatio);
+            const top = Math.max(0, prediction.bbox[1] * yRatio);
+            const width = Math.max(0, prediction.bbox[2] * xRatio);
+            const height = Math.max(0, prediction.bbox[3] * yRatio);
+
+            // check if it is a new region
+            if (regions.length === 0 || !regions.find((region) => region.boundingBox &&
+                    region.boundingBox.left === left &&
+                    region.boundingBox.top === top &&
+                    region.boundingBox.width === width &&
+                    region.boundingBox.height === height)) {
+                regions.push({
+                    id: shortid.generate(),
+                    type: RegionType.Rectangle,
+                    tags: this.state.project.activeLearningSettings.predictTag ? [prediction.class] : [],
+                    boundingBox: {
+                        left,
+                        top,
+                        width,
+                        height,
+                    },
+                    points: [{
+                        x: left,
+                        y: top,
+                    },
+                    {
+                        x: left + width,
+                        y: top,
+                    },
+                    {
+                        x: left + width,
+                        y: top + height,
+                    },
+                    {
+                        x: left,
+                        y: top + height,
+                    }],
+                });
+            }
+        });
+
+        this.canvas.current.addRegionsToAsset(regions);
+        this.canvas.current.addRegionsToCanvasTools(regions);
+
+        const newAsset = { ...this.state.selectedAsset, regions };
+        newAsset.asset.predicted = true;
+        newAsset.asset.state = AssetState.Tagged;
+
+        // Temporary un-comment these two line as calling updateProjectTagsFromAsset
+        // is causing an issue with the dispatcher - to be debugged
+        await this.props.actions.saveAssetMetadata(this.props.project, newAsset);
+        await this.props.actions.saveProject(this.props.project);
+        // Temporary comment this as causing an issue with the dispatcher - to be debugged
+        // await this.updateProjectTagsFromAsset(newAsset, this.props.project, true);
+
+        this.setState({
+            selectedAsset: newAsset,
+        }, async () => {
+            await this.onAssetMetadataChanged(newAsset);
+        });
+    }
```

Comment:
The bulk of this method should be moved up into a common library.  I would expect to see it broken down into the following:

1. Method that takes a `canvas` input and returns an array of `regions`.
2. If project prediction is enabled, run prediction, set `regions` prop on canvas.

Reply:
Partially done, see comments above.  Need to test new refactored method and definitely we need to debug a thing together on adding regions/tags to the project.